tests/unit/test_stub.py

```diff
@@ -0,0 +1,115 @@
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+from tests import unittest
+import mock
+
+from botocore.stub import QueueStubber
+from botocore.exceptions import ParamValidationError, StubResponseError
+from botocore import hooks
+
+
+class TestStubber(unittest.TestCase):
+    def setUp(self):
+        self.event_emitter = hooks.HierarchicalEmitter()
+        self.client = mock.Mock()
+        self.client.meta.events = self.event_emitter
+        self.client.meta.method_to_api_mapping.get.return_value = 'foo'
+        self.stubber = QueueStubber(self.client)
+        self.validate_parameters_mock = mock.Mock()
+        self.validate_parameters_patch = mock.patch(
+            'botocore.stub.validate_parameters', self.validate_parameters_mock)
+        self.validate_parameters_patch.start()
+
+    def tearDown(self):
+        self.validate_parameters_patch.stop()
+
+    def emit_get_response_event(self, model=None, request_dict=None,
+                                signer=None, context=None):
+        if model is None:
+            model = mock.Mock()
+            model.name = 'foo'
+
+        handler, response = self.event_emitter.emit_until_response(
+            event_name='before-call.myservice.foo', model=model,
+            params=request_dict, request_signer=signer, context=context)
+
+        return response
+
+    def test_stubber_registers_events(self):
+        self.event_emitter = mock.Mock()
+        self.client.meta.events = self.event_emitter
+        self.stubber.activate()
+        self.assertTrue(self.event_emitter.register.called)
+
+    def test_stubber_unregisters_events(self):
+        self.event_emitter = mock.Mock()
+        self.client.meta.events = self.event_emitter
+        self.stubber.activate()
+        self.stubber.deactivate()
+        self.assertTrue(self.event_emitter.unregister.called)
+
+    def test_add_response(self):
+        response = {'foo': 'bar'}
+        self.stubber.add_response('foo', response)
+
+        with self.assertRaises(AssertionError):
+            self.stubber.assert_no_pending_responses()
+
+    def test_add_response_fails_when_missing_client_method(self):
+        del self.client.foo
+        with self.assertRaises(ValueError):
+            self.stubber.add_response('foo', {})
+
+    def test_validates_service_response(self):
+        self.stubber.add_response('foo', {})
+        self.assertTrue(self.validate_parameters_mock.called)
+
+    def test_validate_ignores_response_metadata(self):
+        service_response = {'ResponseMetadata': {'foo': 'bar'}}
+        service_model = mock.Mock()
```

Comment:
I'd prefer not to mock service model objects.  Given this is an object we own (and is primarily a data object) I'd prefer to just create the models directly.

We've seen these mocks quickly get out of date as the corresponding APIs they're mocking change.